File: example03adapters/src/main/java/foo/bar/example/foreadapters/feature/playlist/Track.java
package foo.bar.example.foreadapters.feature.playlist;

import android.support.annotation.ColorRes;

/**
 *
 */
public class Track {

    @ColorRes
    private final int colourResource;
    private static final int MIN_PLAYS_REQUESTED = 1;
    private static final int MAX_PLAYS_REQUESTED = 9;
    private int numberOfPlaysRequested = MIN_PLAYS_REQUESTED;

    public Track(@ColorRes int colourResource) {
        this.colourResource = colourResource;
    }

    public int getColourResource() {
        return colourResource;
    }

    public int getNumberOfPlaysRequested() {
        return numberOfPlaysRequested;
    }

    // these aren't full observable models,
    // we want the playlist to manage any changes
    // so we don't make this method public
    protected void increasePlaysRequested() {
        if (canIncreasePlays()){
            numberOfPlaysRequested++;
        }
    }

    // these aren't full observable models,
    // we want the playlist to manage any changes
    // so we don't make this method public
    protected void decreasePlaysRequested() {
        if (canDecreasePlays()){
            numberOfPlaysRequested--;
        }
    }

    public boolean canIncreasePlays(){
        return numberOfPlaysRequested<MAX_PLAYS_REQUESTED;
    }

    public boolean canDecreasePlays(){
        return numberOfPlaysRequested>MIN_PLAYS_REQUESTED;
    }
}

File: example-kt-03adapters/src/main/java/foo/bar/example/foreadapterskt/feature/playlist/Track.kt
package foo.bar.example.foreadapterskt.feature.playlist

import androidx.annotation.ColorRes
import co.early.fore.adapters.DiffComparator


class Track(
        @param:ColorRes @field:ColorRes
        val colourResource: Int,
        val id: Long,
        playsRequested: Int? = null
) : DiffComparator<Track> {

    var numberOfPlaysRequested: Int
        private set

    init {
        numberOfPlaysRequested = playsRequested?.let {
            it
        } ?: MIN_PLAYS_REQUESTED

    }

    fun increasePlaysRequested() {
        if (canIncreasePlays()) {
            numberOfPlaysRequested++
        }
    }

    fun decreasePlaysRequested() {
        if (canDecreasePlays()) {
            numberOfPlaysRequested--
        }
    }

    fun canIncreasePlays(): Boolean {
        return numberOfPlaysRequested < MAX_PLAYS_REQUESTED
    }

    fun canDecreasePlays(): Boolean {
        return numberOfPlaysRequested > MIN_PLAYS_REQUESTED
    }

    companion object {
        private const val MIN_PLAYS_REQUESTED = 1
        private const val MAX_PLAYS_REQUESTED = 9
    }

    override fun itemsTheSame(other: Track?): Boolean {
        return if (other != null) {
            this.id == other.id
        } else false
    }

    override fun contentsTheSame(other: Track?): Boolean {
        return if (other != null) {

            this.colourResource.toString() + "kjhkjhkhkhkhjhkjh" == other.colourResource.toString() + "kjhkjhkhkhkhjhkjh"
                    &&

            this.numberOfPlaysRequested == other.numberOfPlaysRequested
                    && this.colourResource == other.colourResource
        } else false
    }

    fun copy(): Track {
        return Track(
                colourResource,
                id,
                numberOfPlaysRequested
        )
    }
}
